File: js/241_DifferentWaystoAddParentheses.js
/**
 * @param {string} input
 * @return {number[]}
 */
var diffWaysToCompute = function(input) {
    var ret = [];
    for(var i = 0; i < input.length; ++i){
        var c = input[i];
        if(c === '+' || c === '-' || c === '*'){
            var left = diffWaysToCompute(input.substring(0,i));
            var right = diffWaysToCompute(input.substring(i+1));
            for(var m in left){
                for(var n in right){
                    ret.push(c === '+' ? left[m] + right[n] : c === '-' ? left[m] - right[n] : left[m] * right[n]);
                }
            }
        }
    }
    
    if(ret.length < 1 && input.length > 0){
        ret.push(Number.parseInt(input));
    }
    return ret;
};

File: cpp/241_DifferentWaystoAddParentheses.cpp
class Solution {
public:
    vector<int> diffWaysToCompute(string input) {
        vector<int> ret;
        for(int i = 0; i < input.length(); ++i){
            char c = input[i];
            if(c == '+' || c == '*' || c == '-'){
                vector<int> leftRet = diffWaysToCompute(input.substr(0,i));
                vector<int> rightRet = diffWaysToCompute(input.substr(i+1));
                for(auto a : leftRet){
                    for(auto b : rightRet){
                        ret.push_back( c == '+' ? a+b : c == '-' ? a-b : a*b);
                    }
                }
            }
        }
        if(ret.empty()){
            ret.push_back(std::stoi(input));
        }
        return ret;
    }
};